design-proposals/vm-memory-dump.md

```diff
@@ -0,0 +1,99 @@
+# Overview
+Proposal for taking VM memory dump for analysis purposes.
+
+## Motivation
+One of the methods of troubleshooting guest virtual machines is getting a memory dump and analyzing it. As this is a standard capability in most VM management systems, it is also required for kubevirt’s VMs.
+
+## Goals
+Have a mechanism to get a VM memory dump.
+
+## Non Goals
+Memory dump is not a restorable output and currently will not be a part of the VM snapshot.
+
+## User Stories
+As a Kubevirt user I would like to get a memory dump of a running VM so that I can later inspect its memory.
+
+## Repos
+[KubeVirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+To trigger a memory dump a virtctl memoryDump command will be added.
+The command will either use an existing PVC or create a new PVC on demand. This PVC will be bound to the VM and will appear in the VM volumes spec with a fitting status.
+During the memory dump process this PVC will be mounted to the virt-launcher pod. After the mounting the guest memory will be dumped to that pvc and eventually it will be unmounted from the virt launcher.
+The PVC will remain bound - in the VM spec. It is the users' responsibility to export/use the memory dump results before reusing the PVC for a new memory dump. As long as the PVC is bound as a memory dump "container" each memory dump command will overwrite the previous content.
+It will be possible to track the last memory dump in the VM status with a new memory dump status that will be added there.
+It will be possible to unbound the PVC and then do a memory dump to another PVC that will be bound instead.
+
+## API
+
+### Get memory dump
+#### trigger memory dump
+Run a virtctl command to get a memory dump. This new API will look as follows:
+
+`$ virtctl memory-dump my-vm --volume-name=memory-pvc`
+
+In this case `memory-pvc` should already exist and be of a size big enough to contain the memory dump. A check will be made to make sure.
+It will be possible to ask for a PVC to be created with `--create` flag. In such case the required size will be calculated.
+
+##### The process
+The trigger of memory dump will call a VM subresource.
+The VM subresource endpoint will recieve a request containing the pvc name the user want to dump the memory to.
```

Comment:
Can you please extend on the VM subresource ep?

Reply:
what more should be extended on? I explained about what it will update and trigger what is missing?

Reply:
simply how it looks in the VM CR status perspective

Reply:
Not sure I understand the endpoint is not shown in the VM CR, its a a path with an ending /memorydump that redirects the request and eventually updates the VM CR status with the MemoryDumpStatus with values of volume name and binding status which I wrote about and shown the struct

Reply:
Ok, that explanation made it much clearer to me, so consider adding it.